### DISCUSSION TITLE
Bump MSRV to 1.48?

### DISCUSSION THREAD

**Discussion by Kixunil** [rust-bitcoin/rust-bitcoin#1329] [General]
I'd like to propose bumping MSRV again to 1.48. This version is 2 years old now and is available in Debian stable. (Interestingly Buster, containing 1.41 is now EOL) I believe the benefits of this are significant mainly thanks to a bunch of `core` traits being implemented on all arrays. We have several arrays above 32 items and so could remove a bunch of manual trait impls and improve APIs. E.g. we could require `Hash::Inner: AsRef<[u8]>` which would improve generic processing of hashes (see https://github.com/rust-bitcoin/rust-bitcoin/pull/1321#discussion_r999157710), I think we could have a better `ToHex` trait.

Other things I find interesting:

* better `const` support
* `#[track_caller]` - we wouldn't have to conditionally enable it as I suggested.
* the MSRV for `no_std` becomes the same removing a bunch of complexity in docs and maybe also CI

Alternatively, 1.47 provides most (maybe all) of the desired benefits and is already MSRV for `no_std`. I find 1.48 more practical but if anyone needs 1.47 I'm OK with that.

**Comment by apoelstra** [comment]
These are compelling reasons. (weak) concept ACK from me.

**Reply by apoelstra** [reply]
We would also want to decide how to do this -- last time we did a major release for all projects across the rust-bitcoin ecosystem, which I liked because it didn't break any of my projects, though it was a significant amount of work and @Kixunil had an argument that this was actually a bad way to do it.

Downstream users who get broken are of course able to use `cargo update -p` the same way we do, if they need to preserve MSRV.

**Reply by Kixunil** [reply]
If we're going to do this then I'd like to take advantage of it in APIs, mainly the `bitcoin_hashes::Hash` trait which is a breaking change (adding a bound to an associated type publicly-implementable trait) and so we will have to change it anyway. Just bundle it with other breaking changes I guess.

**Comment by sanket1729** [comment]
I am in favor of 1.51 which would give us const generics and Array::IntoIter. But we can also do that later and the informal 2-year rule would put 1.51 on Feb end 2023

**Reply by apoelstra** [reply]
I'd prefer to maintain all of these as hard rules:
* 2 years (ok, we wait til March, no big deal)
* mrustc (it's at 1.54 apparently, nice!)
* debian stable (1.48 apparently)

So currently I'm a nack on 1.51.

**Reply by Kixunil** [reply]
`IntoIterator` on array is very compelling but Debian stable is a hard requirement from me as well. It coud happen that some new version lands in Debian (did happen with 1.36 -> 1.41.1 in Buster IIRC) in which case I'd ACK.

**Comment by TheBlueMatt** [comment]
Just because its soon-ish anyway, I'd suggest we do this after Debian bookworm releases, which puts the 1.41 people on a then-ancient version of Debian. That should be around 2023-04 or so, depending on how long a few things take. I only suggest this because none of the reasons here seem super pressing to me - I agree they're all nice, but if they wait a few releases it doesn't seem like all that much difference? If y'all think this is pressing I'd be okay with seeing it now-ish.

**Reply by Kixunil** [reply]
Since I'd like to take advantage of it in API we at least shouldn't do it after we have 1.0. If it takes some time it's probably OK. Maybe I will try make a demo PR showing how much it'd change if I get some free time.

> That should be around 2023-04

Wow! I though it'll be around summer. Suddenly Debian releases don't feel slow. :D

**Reply by TheBlueMatt** [reply]
Heh, I think it varies - more bugs -> summer, less bugs -> April/May

**Comment by tcharding** [comment]
Even if MSRV bump was the next thing to do after current thing (re-org / crate smashing) it would, in my opinion, not happen till March - the crate smashing PRs take ages to merge because they all conflict plus there is going to be an epic amount of discussion required to work out what goes where. We've only just scratched the surface and its been going on for a while already.

**Reply by Kixunil** [reply]
Indeed, if we decide today to bump MSRV the release will be out much later. I don't expect that crazy changes as the last time because we're jumping versions with fewer language changes. I'd personally mostly focus on array APIs and do the rest later.

**Comment by Kixunil** [comment]
Another possibility: bump to 1.48 ~now and to 1.51 somewhat soon after bookworm release. 1.51 contains much more const generics features. Maybe we could even change the `Hash` trait to remove `Inner` and write `[u8; Self::LEN]` instead.

**Reply by TheBlueMatt** [reply]
I don't think its reasonable to assume debian stable a week after its released, sadly.

**Reply by Kixunil** [reply]
Yeah, that's why I write *somewhat* soon (two months?) and also would like to release a proper version before bumping MSRV so that people can stay on an old one for a bit.

**Reply by TheBlueMatt** [reply]
You have an incredibly optimistic take on how quickly Debian users upgrade to the latest stable after its released :p (given oldstable is maintained for years).

**Reply by Kixunil** [reply]
I don't, I lag behind too. :)

My idea is to have something usable for them - a version with fewer features maybe while moving on. We'd still backport bug fixes and I'd even consider accepting PRs backporting features that are unachievable in consumer code.

**Reply by TheBlueMatt** [reply]
I'm very much against the "well, we did a release before we bumped the MSRV" approach - ultimately most of the rust-bitcoin users are downstream-of-downstream, and sometimes downstream-of-downstream-of-downstream. Ultimately the people who care about the MSRV (those building rust-bitcoin) are rarely the ones who decide when to upgrade the rust-bitcoin dependency :(.

Obviously we should, at a minimum, do a release before we bump the MSRV, but I don't think that's a mitigating factor for when we decide to do an MSRV bump.

**Reply by Kixunil** [reply]
> Ultimately the people who care about the MSRV (those building rust-bitcoin) are rarely the ones who decide when to upgrade the rust-bitcoin dependency

I think this affects other dependencies, so it's not our responsibility if a crate in the middle of the tree does something unfortunate.

But I'm fine with some conservative values for our magic constants.

**Reply by TheBlueMatt** [reply]
I'm really confused what your point is here? Yes, its not our responsibility to make sure crates which depend on rust-bitcoin don't rely on a super recent rustc, of course, but dependent crates (eg LDK, BDK, etc) are going to want to rely on the latest rust-bitcoin, usually for good reason. Folks downstream of LDK/BDK are *also* going to want to rely on the latest version of LDK/BDK, usually for good reason. The *users* of those downstream projects are the ones who ultimately care about the MSRV, they're the ones building. Sure, we can't ensure they get a decent MSRV, but we can ensure that its *possible* for them to get a decent MSRV while still running the latest software. If we took the stance that users who want an MSRV can just run old software then we shouldn't bother with an MSRV at all - users can just go back and use two-year-old rust-bitcoin if they want two-year-old rustc.

**Reply by Kixunil** [reply]
My point is that the balance between compatibility and new compiler features *may* be somewhere else. I think backporting fixes is a middle ground. If downstream crates need to use many new features, then it seems `rust-bitcoin` is still very far from feature-complete and I'd love to know which important things are missing.

Another approach that may make sense is to define MSRV by taking into account planned release date of `master` branch. For instance if we know we will release no sooner than in 6 months and that a certain MSRV is unacceptable right now but would be at the time of release we bump `master` immediately if there are important language features.

Examples of important features:

* const generics (1.51)
* weak dependencies in cargo (1.65)
* GAT probably (1.65 IIRC)

**Reply by TheBlueMatt** [reply]
Sure, if the features are really important in usability of the crate I'd agree. But, frankly, I can't name a rust feature that is *that* important in usability for downstream crates. There's a few that let you clean up code and make things more readable, sure, but only a bit, really, and even then they don't change how easy it is for downstream users to use your crate.

Maybe the one big exception being async/await, but I don't think that applies here.

The obsession in the rust community with using brand new features which are helpful but not massive wins while pretending there's no way they could possibly write code without the feature from yesterday irks me to no end.

**Reply by Kixunil** [reply]
I believe all three I listed are pretty massive. Ironically, async/await are currently pretty much unusable and will be in some future release.

**Reply by TheBlueMatt** [reply]
I don't see how any of the above listed items are "massive". Nice, sure, but, for example, removing some conditional compilation, docs, and a few CI jobs isn't a "massive" win. Same goes for const, some things which were already going to be const-ified by LLVM will now be explicitly const, maybe even a few things will become const that LLVM didn't figure out. Better performance in some cases, cool. Not "massive".

**Reply by apoelstra** [reply]
I philosophically agree with @Kixunil here, but I'd like to see specific examples of how we can use these things in rust-bitcoin. My suspicion is that it will take months or even years of iteration to figure out what the best way to use these new features us. It's also not obvious to me that in rust-bitcoin (as opposed to rust-miniscript, say, where the lack of GATs frequently makes me consider abandoning the project and just switching to Haskell) there are opportunities to use these feautres. I'll bet it takes some months or even years to figure out how to effectively use GATs here.

We definitely need weak dependencies, but I understand Matt's POV that we've managed to hack around them so far, and it's not the end of the world to hack around them for a few more years. And I feel less bad about having an exponential explosion of cargo features now that I know *eventually* that they'll be able to be removed.

Anyway, I'm chipping in my opinion that these features are massive, but in terms of practical consequence, there's nothing we can do to get them any faster than just waiting several years.

**Reply by TheBlueMatt** [reply]
Yea, maybe to clarify my view - some of these things are massive for crate that have major problems addressed by them. GATs, async, etc all solved major pain points for some crates. I don't buy that they're "major" *for rust-bitcoin*.

**Reply by Kixunil** [reply]
I'm open to admit that in some context some of the above mentioned features may be of medium usefulness rather than high. I'll need to play around to find out which.

Note that for me any feature that allows improving the API in an important but backwards-incompatible way is considered massive. That is because we have to choose between stabilizing later or stabilizing worse API. Note that, at least for me, an unstable crate or a crate with ancient API constructs is about as bad as a crate with very lax MSRV and I suspect many people may perceive it even worse. I was actually already thinking about which MSRV we want to stabilize with. (So far I concluded that I need more information and I have to re-think later.)

> Anyway, I'm chipping in my opinion that these features are massive, but in terms of practical consequence, there's nothing we can do to get them any faster than just waiting several years.

For GAT (which I don't think we need *that* much, if at all, in `bitcoin`) and weak dependencies I 100% agree. For some const generics (1.48) I think we can start *planning* MSRV bump today. We could release 0.30 with MSRV 1.41.1 and then do a quick MSRV-bumping release just for const generics & stuff. Combined together this could significantly speed up stabilization of at least some parts of the crate which would be awesome.

**Reply by TheBlueMatt** [reply]
> Note that, at least for me, an unstable crate or a crate with ancient API constructs is about as bad as a crate with very lax MSRV and I suspect many people may perceive it even worse.

Yea, probably generally true, but I obviously strongly disagree. I still find this general view to be one of the Rust community's biggest shortcomings, and a general embarrassment to Rust, not to mention a common thing that stops people from using Rust.

**Reply by Kixunil** [reply]
Why should Rust be the same as other languages though? Different language, different culture... :man_shrugging: (Although I do find it amusing that people supposedly concerned with security don't understand the value of signed packages...)

**Reply by TheBlueMatt** [reply]
Because there's good reason things were done the way they are in other languages, and the mess of npm should have showed us that that specific culture results in way more pain and suffering than it adds value. The rust community should learn from the mistakes of others, not repeat them because they're new and shiny.

**Reply by Kixunil** [reply]
> Because there's good reason things were done the way they are in other languages

Please, be specific, there are many languages with many different good and bad ideas and some of the "good reasons" may not apply to Rust.

> the mess of npm

Again, please be specific. NPM **is** a mess but from my POV Rust solves almost all of its problems (signing is the missing part).

**Reply by apoelstra** [reply]
>Again, please be specific. NPM is a mess but from my POV Rust solves almost all of its problems (signing is the missing part).

I spent 5 or 10 minutes playing with crev the other day. I think if we got into a habit of using this I'd be **much** more comfortable with dependencies and Rust's behavior. We could maybe have an in-repo script that users can run which pin our dependencies to only ones that have been vetted, or highlight ones that haven't. I also learned about a tool `cargo-supply-chain` which lists everyone with crates access to our deps.

Without crev my perception of cargo is that it's only "better than NPM" because the sanity waterline is higher in Rust-land than in JS-land, not because there's any inherent improvement to the problems of malicious crates, giant unvettable dependency trees, and automatic updates.

I'll open a new discussion with these notes.

**Reply by Kixunil** [reply]
Sure, we need `crev` and as I say, this is missing part in Rust. This discussion was mainly about MSRV-related issues. The stability guarantees of Rust should make it possible to update it faster without significantly increasing the risk but it is true that some distributions do not exploit this (yet?).

**Comment by tcharding** [comment]
So are we willing and able to bump MSRV to 1.48.0 directly after 0.30.0 is released and do a quick 0.31.0? (I did re-read the whole thread but its Friday afternoon and I didn't really get a clear idea of where we stand).

**Reply by Kixunil** [reply]
Didn't re-read but that's exactly what I remember.

**Reply by apoelstra** [reply]
I don't think we need to do a quick 0.31.0. On reflection I think the "do a release just to bump MSRV" process serves no real purpose -- if people want to have a release in which they *only* bump their compiler version but make no other changes, they can use literally any release prior to the MSRV bump.

But yes, let's bump to 1.48.0 immediately after 0.30.0 is released.

**Reply by Kixunil** [reply]
I think "quick" meant to also do related changes?

**Reply by apoelstra** [reply]
I think we can get those changes out at their own pace. Though we can certainly prioritize them if they make our lives easier.

**Comment by tcharding** [comment]
Done in #1729
